### ISSUE TITLE
Selective monitoring of Kubernetes workloads from the resource side for fleet

### ISSUE THREAD

**Issue by mlunadia** [elastic/elastic-agent#614]

**Comment by ChrsMark** [comment]
Maybe here we can leverage https://github.com/elastic/observability-dev/blob/main/docs/dc/integrations/testing/api-kubernetes-automation.md so as Agent to declare packages+policies based on the received hints through the Fleet API.

**Comment by gizas** [comment]
Just to clarify, we would like to investigate whether the Fleet api capabilities can be used from our side to support the autodiscovery feature in managed agent scenario.

The idea is that customer defines an annotation with specific format on k8s resource eg label: redis--> this triggers an update in the agent policy of fleet --> this installs the relevant package as well with assets

So I paste two questions here:

- Can we perform such api requests internally inside kibana (without authentication any sample ?)
- Can you point us an example of triggering function  (based on agent configuration values) to update policies etc?

**Comment by kpollich** [comment]
> Can we perform such api requests internally inside kibana (without authentication any sample ?)

I'm not sure I understand what the actual API request we're asking about here is. Is the `label` value a new field on a package policy? If so we expose various CRUD APIs around package policies that should be usable here, and will include the package installation -> policy compilation -> deployment to running agents steps as needed. 

https://petstore.swagger.io/?url=https://raw.githubusercontent.com/elastic/kibana/main/x-pack/plugins/fleet/common/openapi/bundled.json#/default/create-package-policy

@nchaulet - I think we need to update the `PUT /package_policies` docs to use the new simplified format for `inputs`. 

> Can you point us an example of triggering function (based on agent configuration values) to update policies etc?

What is a "triggering function" and what are "agent configuration values" in this case?

Agent Policy updates (maybe these are the "agent configuration values" here?) go through Fleet's `AgentPolicyService.update` public method here:
https://github.com/elastic/kibana/blob/main/x-pack/plugins/fleet/server/services/agent_policy.ts#L384-L428. 

Fleet will then redeploy the policy following the update:
https://github.com/elastic/kibana/blob/main/x-pack/plugins/fleet/server/services/agent_policy_update.ts#L31-L58

Hopefully this helps clarify. I _think_ the Fleet API should be able to handle the use-cases here, but I'm having trouble completely understanding what's planned here from an implementation standpoint.

**Comment by ChrsMark** [comment]
From my perspective the main question is: Is it a valid approach to have Elastic Agent executing API requests to the Fleet API?

I remember in the past it was not such an approved communication schema but maybe this has now changed (or I remember wrong :)).

In our scenario we would like to evaluate if Elastic Agent can dictate a package addition by performing the proper API request to the Fleet API. Would that communication schema be acceptable?

**Comment by kpollich** [comment]
> In our scenario we would like to evaluate if Elastic Agent can dictate a package addition by performing the proper API request to the Fleet API. Would that communication schema be acceptable?

No I would say this is still not an acceptable communication pattern. Communication like this should be handled through Fleet Server and its `actions` implementation, I think. Alternatively, a distinct API for these specific updates could be implemented in Fleet Server, and agents could consume that API. 

Either way, agents should not directly consume the Fleet API.

**Comment by ChrsMark** [comment]
> > In our scenario we would like to evaluate if Elastic Agent can dictate a package addition by performing the proper API request to the Fleet API. Would that communication schema be acceptable?
> 
> No I would say this is still not an acceptable communication pattern. Communication like this should be handled through Fleet Server and its `actions` implementation, I think. Alternatively, a distinct API for these specific updates could be implemented in Fleet Server, and agents could consume that API.
> 
> Either way, agents should not directly consume the Fleet API.

Makes sense! In that case maybe Agents can direct these information to the Fleet server and then Fleet server can make the decisions etc and declare the addition of a new package into a policy through the Fleet API.

**Comment by gizas** [comment]
@kpollich sorry the link provided was broken: This is the kibanaOpenAPI we were referring to and we had created small poc with get/post commands: https://github.com/elastic/observability-dev/blob/main/docs/infraobs/cloudnative-monitoring/testing/api-kubernetes-automation.md 

Sorry for bad analysis in the pic provided, and forgive my ignorance here:
<img width="1547" alt="Screenshot 2022-09-20 at 4 54 44 PM" src="https://user-images.githubusercontent.com/1685415/191277034-4acf5ad2-624a-442f-983f-3641febafb51.png">

1.Agent will send data to Fleet Server initially
2.Does Fleet Server have the ability now to analyse incoming info and trigger a policy update or consume any API ? No if understand it correctly. So this needs to be implemented?
3.Do we have something on ElasticSearch side that monitors updates of configuration and can trigger API requests? 
What is the way right now that we configure a new policy in Fleet Server?

**Comment by nchaulet** [comment]
> 1.Agent will send data to Fleet Server initially
2.Does Fleet Server have the ability now to analyse incoming info and trigger a policy update or consume any API ? No if understand it correctly. So this needs to be implemented? From discussions having with Nicolas I understood that we dont want to create direct connection between Fleet Server and Fleet.
3.Do we have something on ElasticSearch side that monitors updates of configuration and can trigger API requests?
What is the way right now that we configure a new policy in Fleet Server?

I think we could have something working like 

1. Agent will send data to Fleet Server initially
2. Fleet server will store these data somewhere it's accessible by Kibana (need define what data are send)
3. We could have a background task in Kibana that check for new data, and do the proper policy update

**Comment by ChrsMark** [comment]
What @nchaulet describes here sounds quite close to a PoC we had worked on in the past as part of an [OnWeek project](https://github.com/elastic/observability-dev/issues/2100) (along with @MichaelKatsoulis and @hop-dev).

In this case, all the "intelligence"  was implemented in FleetUI/Kibana as a background task. In this regard, the Kibana/FleetUI acts as a controller which is informed for new data written in an ES index and takes the proper decision accordingly, updating the policy with the proper package/input.

While it sounds sane to me and is already to proven that works, my concern is if you overload Kibana/FleetUI with functionality and how would that scale when it comes to hundreds or thousands events to be handled.

**Comment by gizas** [comment]
So from the descriptions above, I list the following:

1. One issue on **cloudnative team** to implement the "correct" policies to be send when hints are enable
2. Story (for Fleet Team): Fleet-Server or Elastic-Agent (?) acquiring and saving policies to elasticsearch. 
3. Story (for Fleet Team): Fleet (kibana) to have a proxy that will monitor specific fields of the policy--> extract hints data and from those install needed packages and create/update agent policies

I note here the concern of @ChrsMark is this solution is scalable? Do we have any concern? 
Also as long as we dont introduce any new background connection I dont see any security concerns. 

@nchaulet , @kpollich are we ok to such stories above? 
Do you want me to create them for you and add inside relevant information?
I see that 2 stories fall back to your backlog and we can sync on how we can help on those and for the prioritisation of those. What do you think?

**Comment by kpollich** [comment]
@gizas - Yes feel free to create those issues and link here. Thank you.

**Comment by gizas** [comment]
@kpollich , @ChrsMark based on above https://github.com/elastic/elastic-agent/issues/614#issuecomment-1259227520:

New stories created:
1. https://github.com/elastic/elastic-agent/issues/1350
2. https://github.com/elastic/elastic-agent/issues/1349
3. https://github.com/elastic/elastic-agent/issues/1347

Please feel free to update add comments per story. 
#1349 might not needed if we go with direct saving of information from elastic-agent to elasticsearch.

**Comment by gizas** [comment]
@kpollich , @ruflin did some small updates on above issues and added a workflow diagram that might help discussions.

Still there are some things we can discuss like:
- Does Fleet server need to save the agent policy status somewhere or checks and updates are done on the fly?
- Such a feature will add heavy load of checks/ updates on fleet server. Load considerations? How can we tackle them?
- On step 2a where we need to update agent policies can we reuse the templates of https://github.com/elastic/kibana/issues/135624? Will this offload the fleet server transactions?
